src/solver/optimize.cpp

```diff
@@ -1,14 +1,296 @@
+#include <math.h>
+#include <string>  
+#include <iostream> 
+#include <sstream> 
 #include "../components/token.hpp"
 #include "../components/order.hpp"
 #include "../components/amm.hpp"
 #include "./optimize.hpp"
+#include "../../external_libs/json.hpp"
 
-void solve_auction(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms)
+
+// auxiliary function (used temporarily) for computing x^n using the boost library data types
+boost::multiprecision::mpf_float custom_pow(boost::multiprecision::mpz_int x, int n)
+{
+    int i;
+    boost::multiprecision::mpf_float res = 1;
+    int m = (n > 0) ? n : (-n);
+
+    for (i = 0; i < m; i++)
+        res = res * x;
+    if (n > 0)
+        return res;
+    else
+        return (1/res);
+}
+
+// main function that processes a sell order, checks if it can be matched against a single liquidity pool
+// and if so, it also checks whether it gives a strictly better objective
+void process_sell_order(std::vector<Token> &tokens, std::vector<Order> &orders, std::vector<CP_AMM> &amms, int i, ExecutedOrder &best_order)
 {
-    // TO BE COMPLETED
-    int *vertices;
-    int **edges;
+   int j, num_amms;
+    boost::multiprecision::mpz_int x0, y0, x1, y1, x_, y_, x;
+    boost::multiprecision::mpf_float float_x1;
+    boost::multiprecision::mpf_float float_x;
+    boost::multiprecision::mpf_float temp;
+
     
+    x_ = orders[i].buy_amount;
+    y_ = orders[i].sell_amount;
+    num_amms = amms.size();
+
+    for (j = 0; j < num_amms; j++) {
+        x0 = 0;
+        y0 = 0;
+        if (amms[j].sell_token_index == orders[i].sell_token_index)
+            x0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].sell_token_index)
+            x0 = amms[j].buy_reserve_amount;
+        if (amms[j].sell_token_index == orders[i].buy_token_index)
+            y0 = amms[j].sell_reserve_amount;
+        if (amms[j].buy_token_index == orders[i].buy_token_index)
+            y0 = amms[j].buy_reserve_amount;
+
+        if ((x0 <= 0) || (y0 <= 0))
```

Comment:
tip: no need for parenthesis, and can use "or" instead of "||".

Reply:
Coming from C, I am a bit more comfortable with "||" and slightly prefer it to be honest! I did remove the nested parentheses.